### ISSUE TITLE
[Autocomplete] required choice types end up not being required.

### ISSUE THREAD

**Issue by bendavies** [symfony/ux#402]
Hi there,

There seems to be an issue with how tom-select is configured, for required, non-multiple choice types.

Given a form type configuration like

```php
$builder->add('isAttending', ChoiceType::class, [
    'choices' => [
        'Yes' => 'yes',
        'No' => 'no',
        'Maybe' => 'maybe',
    ],
]);
```
You will be left with a select box like:
<img width="408" alt="image" src="https://user-images.githubusercontent.com/625392/180837738-d23c78e7-f036-4b15-8fe7-242564f69273.png">

All well and good, an option is mandatory.

However, convert to `autocomplete` and you're left with:
<img width="405" alt="image" src="https://user-images.githubusercontent.com/625392/181054541-ddbddd8e-dee0-4870-8133-dfad67bb01c1.png">

see the issue? It's clearable because the `clear plugin` [is added](https://github.com/symfony/ux/blob/8e84d49a92b70f99001b8817cd3129a6215175f6/src/Autocomplete/assets/src/controller.ts#L52), so you can be left with no value.

<img width="404" alt="image" src="https://user-images.githubusercontent.com/625392/181055492-3af1ba90-62a3-477e-bc15-325384ed1957.png">

this shouldn't be possible, as it isn't with the simple select box above.

We can try fix fix:
```php
$builder->add('isAttending', ChoiceType::class, [
    'choices' => [
        'Yes' => 'yes',
        'No' => 'no',
        'Maybe' => 'maybe',
    ],
    'autocomplete' => true,
    'required' => true,
]);
```

this has no effect for 2 reasons:
1. the symfony/ux-autocomplete controller still adds the clear plugin (while the [easyadmin tom-select controller](https://github.com/EasyCorp/EasyAdminBundle/blob/579fcb45a3f98e66b2fcf40e01d54cd8ffe05104/assets/js/autocomplete.js#L36-L38) takes `requried` into considertion, from which symfony/ux-autocomplete was inspired by)
2. the easy admin version wouldn't work either, because [symfony removes the required attribute](https://github.com/symfony/symfony/blob/b1b77f366d42a58c1cd0879c605b997d05279d01/src/Symfony/Bridge/Twig/Resources/views/Form/form_div_layout.html.twig#L59-L61) for select boxes when it's impossible not to submit a value.

Number 2 can be "fixed"/overridden like this:
```php
$builder->add('isAttending', ChoiceType::class, [
    'choices' => [
        'Yes' => 'yes',
        'No' => 'no',
        'Maybe' => 'maybe',
    ],
    'attr' => [
        'required' => true,
    ],
    'autocomplete' => true,
]);
```

but that would mean we still have to fix the stimulus controller here to **not** add the clear button if there is a `required` attribute.

my "fix" above (adding `'attr' => ['required' => true]`) is probably too onerous to put on the end user. Maybe the autocomplete form type should do this automatically?

Thanks

**Comment by mario-fehr** [comment]
Duplicate of #379 and fixed with #401

**Comment by bendavies** [comment]
I do not agree that those issues are the same as this one - those are about completely overriding the default plugins setup by `symfony/ux-autocomplete`.

this issue: I think symfony/ux-autocomplete should be more intelligent/correct out of the box with reguards to required `select` elements.

**Comment by maartendekeizer** [comment]
@bendavies Agree, not the same but related.
I think best way is to only add the clear button when the select is not required. Another option is that we never set plugins automaticly but I think that gives bad DX

**Comment by weaverryan** [comment]
Hi there!

Really excellent issue and explanation of the problems Ben - thanks for that. As Ben mentioned, the `required` "form view variable" is not always printed out as a `required` attribute on the `select` element, which is why the current solution of looking for that attribute in JavaScript doesn't work. Could we, instead, read the `required` variable (`$view->vars['required']`) in the form type extension - https://github.com/symfony/ux/blob/8e84d49a92b70f99001b8817cd3129a6215175f6/src/Autocomplete/src/Form/AutocompleteChoiceTypeExtension.php#L42 - and then use it to pass in a new "value" to the Stimulus controller?

**Comment by bendavies** [comment]
yep, that looks doable. we just need to pass a new required `attr` which will make it into the template

**Comment by LucileDT** [comment]
@bendavies do you have any news? :eyes:

**Comment by bendavies** [comment]
forgot about this. i'll take a look again

**Comment by Qronicle** [comment]
This is still an issue as far as I can tell. How do we proceed to get the PR merged?

Although thinking about this some more, I feel like the "clearability" of a select is related more to whether there is a placeholder or not. (At least that's how I have always implemented my own js enhanced select elements.)\
You can for example have a required select with an empty/placeholder option that forces the user to manually select something. In this case it would still make sense to show the clear button to revert back to the initial, unset state.

I would have to dive into the code to see how the placeholder is deduced, although I wouldn't be opposed to the solution proposed in the PR, as it would still be a major improvement over what there is now.

Edit: I made an attempt at a pull request that implements the placeholder check

**Comment by Bartheyrman22** [comment]
I believe the placeholder is indeed doing the trick.
We should always add a placeholder in my opinion. Regardless the required-state. Does this make sense?

```php
// Symfony\UX\Autocomplete\Form\AutocompleteChoiceTypeExtension

// Directly after line 102 ( at the end of finishView)
// Add placeholder to add Tomselect required validation.
if ($options['required'] === true && $options['placeholder'] === null) {
    $view->vars['placeholder'] = '';
}
```

This is also working fine:
```php
$builder
    ->add('testElement', ChoiceType::class, [
        'autocomplete' => true,
        'choices' => $choices,
        'placeholder' => '',
        'required' => true,
    ])
```
